HomeKit: add Thread device (#27892)
home-assistant/home-assistant.github.io: source/_integrations/homekit_controller.markdown

HomeKit controller will poll your devices, but it will also automatically enable push updates for accessories that support it.

## Adding a HomeKit device through Bluetooth

> You can add a HomeKit device to Home Assistant via Bluetooth.

Reviewer: ```suggestion
You can add a HomeKit device to Home Assistant via [Bluetooth](/integrations/bluetooth).
```

### Prerequisites

> - If your Home Assistant instance does not natively support Bluetooth, use a Bluetooth proxy. 

Reviewer: ```suggestion
- If your Home Assistant instance does not natively support Bluetooth, use an ESPHome Bluetooth proxy. 
```

  - A proxy can also be helpful if your Home Assistant device is too far away from the device you are trying to pair.
- If your HomeKit device has been used with Thread before, or is still paired with iOS, reset the device.
   - HomeKit devices can only be paired to a single controller at once. 

>    - If it has been in a Thread network before, the device might remember the Thread credentials of a different network. A reset fixes that.

Reviewer: ```suggestion
   - If it has been in a Thread network before, the device might remember the Thread credentials of a different network. A reset makes sure the device is not connected to any Thread network.
```

## To add a HomeKit device through Bluetooth

1. Power up your HomeKit device.
   - If you have Bluetooth enabled, the device should be discovered under **{% my integrations title="Settings > Devices & Services" %}**.
1. On the HomeKit integration, select **Configure**.
     ![HomeKit integration](/images/integrations/homekit_controller/homekit_controller_add_01.png)

> 1. To pair the device, enter the HomeKit pairing code. The code is on the device itself or the packaging. 

Reviewer: Information from the step below:

```suggestion
1. To pair the device, enter the HomeKit pairing code. The code is on the device itself or the packaging. 
   - To pair a battery-powered device, you may need to press a button on the device to wake it.
   - Bluetooth devices may take significantly longer to pair than IP devices.
```

   - Select **Finish**. 
1. The HomeKit device now automatically discovers other Bluetooth devices (if the [Bluetooth](/integrations/bluetooth) integration is enabled).  
   - Bluetooth devices may take significantly longer to pair than IP devices. 

>    - To pair a battery-powered device, you may need to press a button on the device to wake it.

Reviewer: This doesn't belong here anymore, as we have the Bluetooth integration as a prerequisite.

```suggestion
```

## Adding a HomeKit device to a Thread network via Home Assistant

There are two methods to add a HomeKit device to a Thread network:

- via Home Assistant
- via [Apple Thread border router](#adding-a-homekit-device-to-home-assistant-via-apple-thread-border-router)

This section describes how to add it via Home Assistant.

### Prerequisites

- A HomeKit device which supports Thread. This is indicated by the Thread label on the packaging.

> - Make sure HomeKit controller has been [joined using Bluetooth](#adding-a-homekit-device-through-bluetooth). 

Reviewer: ```suggestion
- Make sure the HomeKit device has been [joined using Bluetooth](#adding-a-homekit-device-through-bluetooth). 
```

- **Thread network**: In order to use HomeKit over Thread, you need a working border router. 

>     - Make sure your Home Assistant device is on the same network (LAN) as the border router.

Reviewer: ```suggestion
    - Make sure your Home Assistant device is on the same network (LAN) as the border router.
    - Make sure the Thread network you'd like to use is known by Home Assistant and marked as **Preferred network** in the Thread configuration.
```

    - If you have a Home Assistant Yellow or SkyConnect, you can enable multiprotocol to set up an Open Thread border router and with that a Thread network. 
      - Documentation on [enabling multiprotocol on Yellow](https://yellow.home-assistant.io/guides/enable-multiprotocol/)
      - Documentation on [enabling multiprotocol on SkyConnect](https://skyconnect.home-assistant.io/procedures/enable-multiprotocol/)

> ### To add a HomeKit  device to a Thread network via Home Assistant

Reviewer: ```suggestion
### To add a HomeKit device to a Thread network via Home Assistant
```

   - That's it. Your HomeKit device now communicates via Thread.

## Adding a HomeKit device to Home Assistant via Apple Thread border router

There are two methods to add a HomeKit device to a Thread network:

- via [Home Assistant](#adding-a-homekit-device-to-a-thread-network-via-home-assistant)
- via Apple Thread border router

This section describes how to add a HomeKit device using an Apple Thread border router device such as a HomePod mini.

### Prerequisites

- An Apple device that can act as a Thread border router, such as a HomePod mini.
- A HomeKit device which supports Thread. This is indicated by the Thread label on the packaging.
- Make sure your Home Assistant instance is on the same network (LAN) as the border router.
- Make sure the HomeKit device has been paired in the Apple Home app (using an iOS app).

### To add a HomeKit device to Home Assistant via Apple Thread border router

1. Unpair the HomeKit device from the Apple Home app. Don't reset the device.
   - This leaves the Thread network details on the HomeKit device.
1. The device will be automatically discovered by the HomeKit controller integration in Home Assistant. 

>    - It will appear as a discovered device over Thread.

Reviewer: Just like Bluetooth, you'll have to click configure here. Then it will be added.

Author: Thank you for the catch! Added missing steps.

   - HomeKit devices can only be paired to a single controller at once.
   - If it has been in a Thread network before, the device might remember the Thread credentials of a different network. A reset makes sure the device is not connected to any Thread network.

## To add a HomeKit device through Bluetooth

1. Power up your HomeKit device.
   - If you have Bluetooth enabled, the device should be discovered under **{% my integrations title="Settings > Devices & Services" %}**.
1. Under **{% my integrations title="Settings > Devices & Services" %}**, on the HomeKit integration, select **Configure**.
     
     ![HomeKit integration](/images/integrations/homekit_controller/homekit_controller_add_01.png)
1. To pair the device, enter the HomeKit pairing code. The code is on the device itself or the packaging. 
   - To pair a battery-powered device, you may need to press a button on the device to wake it.
   - Bluetooth devices may take significantly longer to pair than IP devices.
   - Add the device to a room and **Finish**.

## Adding a HomeKit device to a Thread network via Home Assistant

There are two methods to add a HomeKit device to a Thread network:

> - via Home Assistant

Reviewer: "via Home Assistant" on its own is a bit unclear to me. I'd describe it a bit more. The title can still be "via Home Assistant" only.

```suggestion
- via Home Assistant's preferred Thread network
```

- via [Apple Thread border router](#adding-a-homekit-device-to-home-assistant-via-apple-thread-border-router)

> This section describes how to add it via Home Assistant.

Reviewer: ```suggestion
This section describes how to add it via Home Assistant's preferred Thread network.
```

      - Documentation on [enabling multiprotocol on Yellow](https://yellow.home-assistant.io/guides/enable-multiprotocol/)
      - Documentation on [enabling multiprotocol on SkyConnect](https://skyconnect.home-assistant.io/procedures/enable-multiprotocol/)

### To add a HomeKit device to a Thread network via Home Assistant

1. To open the device configuration page, on the **HomeKit** integration, select the **device**.
1. Under **Diagnostic**, you can see the **Thread Status** as **Disabled**.
    ![Device configuration page](/images/integrations/homekit_controller/homekit_controller_add_02.png)
1. To enable Thread, under **Configuration**, select **Press**. This will provision the preferred Thread credentials.
   - The status has now changed: 
     - Depending on the device type, the mesh size and health, the Thread status can be **Child**, **Router**, or **Leader**. 
     ![Thread status](/images/integrations/homekit_controller/homekit_controller_add_02.png)
   - That's it. Your HomeKit device now communicates via Thread.

## Adding a HomeKit device to Home Assistant via Apple Thread border router

There are two methods to add a HomeKit device to a Thread network:

> - via [Home Assistant](#adding-a-homekit-device-to-a-thread-network-via-home-assistant)

Reviewer: As above

```suggestion
- via [Home Assistant's preferred Thread network](#adding-a-homekit-device-to-a-thread-network-via-home-assistant)
```

- via Apple Thread border router

This section describes how to add a HomeKit device using an Apple Thread border router device such as a HomePod mini.

### Prerequisites

- An Apple device that can act as a Thread border router, such as a HomePod mini.
- A HomeKit device which supports Thread. This is indicated by the Thread label on the packaging.
- Make sure your Home Assistant instance is on the same network (LAN) as the border router.

> - Make sure the HomeKit device has been paired in the Apple Home app (using an iOS app).

Reviewer: ```suggestion
- Make sure the HomeKit device has been paired in the Apple Home app (using the iOS Home app).
```

HomeKit Controller will automatically discover Bluetooth devices once the [Bluetooth](/integrations/bluetooth) integration is enabled and functional. Bluetooth devices may take significantly longer to pair than IP devices. Battery-powered devices may require pressing a button on the device to wake it before pairing can be successful.

## Thread device support

In order to use HomeKit over Thread you need a working border router (like a HomePod mini). Your Home Assistant instance will need to be on the same VLAN as the border router.

HomeKit Controller will automatically discover supported Thread devices using the [Zeroconf](/integrations/zeroconf) integration. Battery powered devices may go to sleep and require a button pressing to wake them up before pairing works.

> In order to provision a Thread device onto a mesh network using a HomePod as a border router, first pair it with an iOS device. Then unpair it from the Home app. Don't reset it. This will leave the Thread network details on the device. Home Assistant can now pair with it over Thread.

Reviewer 2: We should try to retain this information, though it should be reworded as it is written as though it is the only configuration procedure which is no longer true. It should probably come after the new procedure as it is a special case.

If you use this original procedure, Apple border routers are absolutely supported. What happens is iOS configures them to join the thread network rather than HA, so HA doesn't need to know the thread credentials for it to work.

One advantage it has over the new procedure is that it is easier to move your iPhone to a device that is screwed to the wall (like an Eve blind) than it is your HA instance. Given bluetooth range limitations (especially for people that haven't invested in a long range bluetooth dongle), this might be the original process is actually better for some users.

> In order to provision a Thread device onto a mesh network using a HomePod as a border router, first pair it with an iOS device. Then unpair it from the Home app. Don't reset it. This will leave the Thread network details on the device. Home Assistant can now pair with it over Thread.

Reviewer: This method is now kinda completely missing.

Since currently we can't import HomeKit network credentials to HA, this is the only way to add a HomeKit device through Apple Border Routers to HA.

Under the "Thread device support" section we need two sections:

1. Provision Thread device through Home Assistant (essentially what you've added now)
2. Provision Thread device through Apple Home app (currently documented method)

> ## Adding a HomeKit controller device to a Thread network

Reviewer: This is somewhat confusing. Since this is already in the HomeKit Controller integration, I'd go for

```suggestion
## Adding a HomeKit device to a Thread network
```

> * **Bluetooth support**: HomeKit devices are joined using Bluetooth. If your Home Assistant server device does not support Bluetooth, you can use a Bluetooth proxy.

Reviewer 2: Is it worth calling out the range specifically here? Something like this?

> If your Home Assistant device is too far away from the device you are trying to pair, you can use a bluetooth proxy.

I don't know the brands but I do know some bluetooth/thread locks actually underpower their bluetooth raido as some kind of security feature. We've had reports that people had to put their HA Pi withing 12" for it to work.

> * **Bluetooth support**: HomeKit devices are joined using Bluetooth. If your Home Assistant server device does not support Bluetooth, you can use a Bluetooth proxy.

Reviewer: There is already a "Bluetooth device support" section above.

Since all HomeKit devices which support Thread do support Bluetooth as well, and a device needs to be paired through Bluetooth first, I'd just reference to that section.

```suggestion
* **Bluetooth support**: The HomeKit device needs to be already connected through Bluetooth (see above).
```

>     * Make sure your Home Assistant device is on the same network (LAN) as the border router. 

Reviewer 2: Should we call out common mistakes here? Or reference a shared thread page for that? (E.g. Cross VLAN not supported, IPv6 must be set to auto in HAOS, i can think of more if this is the right place for that information)

Author: Let's get the procedures done first. Then we can add troubleshooting items in a separate PR.

    * If you have a Home Assistant Yellow or SkyConnect, you can enable multiprotocol to set up an Open Thread border router and with that a Thread network. 
      * Documentation on [enabling multiprotocol on Yellow](https://yellow.home-assistant.io/guides/enable-multiprotocol/)
      * Documentation on [enabling multiprotocol on SkyConnect](https://skyconnect.home-assistant.io/procedures/enable-multiprotocol/)

>     * Other devices, such as an Apple HomePod mini or Apple TV, cannot currently be used as border routers. They do not communicate their Thread credentials an can therefore not be integrated.

Reviewer 2: They can using the original pairing process so that iOS shares thread credentials for the Apple thread network with the HomeKit devices for us, then they pair to HA.

### To add a HomeKit controller device to a Thread network

> 1. Power up your HomeKit controller device.

Reviewer 2: For this pairing procedure, it is probably a good idea to encourage them to do a full reset if the device has previously been used with thread. Otherwise the device might remember the thread credentials of a different network, and this process might not work.

I think its mentioned elsewhere on this page, but might be worth stressing that if a device isn't discovered, make sure its not still paired to iOS. HomeKit devices can only be paired to a single controller at once. A full reset should avoid this case too.

   * If you have Bluetooth enabled, it should be discovered under **{% my integrations title="Settings > Devices & Services" %}**.
1. On the HomeKit integration, select **Configure**.
1. To pair the device, enter the HomeKit pairing code. The code is on the device itself or the packaging. 

>    * Select **Finish**. 

Reviewer: Since we make Bluetooth connection a prerequisite, this won't be necessary.

```suggestion
```

We could move this steps to the Bluetooth section though :thinking:

1. To open the device configuration page, on the integration, select the **device**.
1. Under **Diagnostic**, you can see the **Thread Status** as **Disabled**.
    ![Device configuration page](/images/integrations/homekit_controller/homekit_controller_add_02.png)
1. To enable Thread, under **Configuration**, select **Press**. This will provision the preferred Thread credentials.

>    * The status has now changed to **Child**. 

Reviewer 2: Depending on the device and the mesh size and health, this can also be "Router" and i think also "Leader".

     ![Thread status](/images/integrations/homekit_controller/homekit_controller_add_02.png)
   * That's it. Your controller device now communicates via Thread.

>    * Your HomeKit controller automatically discovers supported Thread devices using the [Zeroconf](/integrations/zeroconf) integration. Battery powered devices may go to sleep. To pair them, you may need to press a button to wake them up.

Reviewer 2: In my own testing, this really only seems to apply to devices that are on a thread network but unpaired. Pairing over bluetooth first seems to avoid it, so we can probably drop that comment here.

>    * Your HomeKit controller automatically discovers supported Thread devices using the [Zeroconf](/integrations/zeroconf) integration. Battery powered devices may go to sleep. To pair them, you may need to press a button to wake them up.

Reviewer: This is related to the other method to on-board a Thread device, so it doesn't belong here.

```suggestion
```
